File: algorithm/leetcode/82_Remove_Duplicates_from_Sorted_List_II.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function(head) {
    if(head === null && head.next === null) return head;
    let prev = new ListNode(0);
    prev.next = head;
    let start = prev;
    let curr = head;
    while(curr) {
        while(curr.next && curr.val === curr.next.val) {
            curr = curr.next
        }
        if(start.next == curr) {
            start = start.next
        }else {
            start.next = curr.next
        }
        curr = curr.next
    }
    return prev.next
};

File: algorithm/leetcode/82_Remove_Duplicates_from_Sorted_List_II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode prev = new ListNode(0);
        prev.next = head;
        ListNode start = prev;
        ListNode curr = head;
        while(curr != null) {
            while(curr.next != null && curr.val == curr.next.val) {
                curr = curr.next;
            }
            if(start.next == curr) {
                start = start.next;
            }else {
                start.next = curr.next;
            }
            curr = curr.next;
        }
        return prev.next;
    }
}